### PR TITLE
Submit assignment-2

### PR DESCRIPTION
**Pull request by WangYP-2022**
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
Implementing four model pipelines with two preprocessors and two regressors, including hyperparameter tuning, evaluation, SHAP explanations, and model export.
## What did you learn from the changes you have made?
MAE is preferred over R² for prediction tasks, and tree-based models with proper preprocessing outperform linear baselines on non-linear data.
## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?
Log-transforming the target variable or using XGBoost/RandomForest instead of GradientBoosting for faster computation.
## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?
Handling the right-skewed target distribution and selecting appropriate metrics, which I addressed by using MAE and tree-based models robust to skewness.
## How were these changes tested?
5-fold cross-validation during GridSearchCV and hold-out test set evaluation with MAE, RMSE, and R² metrics.
## A reference to a related issue in your repository (if applicable)
N/A
## Checklist
- [x] I can confirm that my changes are working as intended

### CODE REVIEW

**Review by Dmytro-Bonislavskyi — COMMENTED** [review]
Hi @WangYP-2022  Well done!

**Comment by Dmytro-Bonislavskyi** [comment]
@WangYP-2022 You provided explanation part for tree-based model, just in case here is an option for Linear models too (A and B here):
```
explainer = shap.LinearExplainer(best_model.best_estimator_.named_steps['regressor'],  X_train_transformed)
```
